src/execution.jl

```diff
@@ -356,92 +423,51 @@ The output of this function is automatically cached, i.e. you can simply call `c
-            image = asm
-            # linking the device runtime library requires use of the CUDA linker,
-            # which in turn switches compilation to device relocatable code (-rdc) mode.
-            #
-            # even if not doing any actual calls that need -rdc (i.e., calls to the runtime
-            # library), this significantly hurts performance, so don't do it unconditionally
-            intrinsic_fns = ["vprintf", "malloc", "free", "__assertfail",
-                            "__nvvm_reflect" #= TODO: should have been optimized away =#]
-            if !isempty(setdiff(undefined_fns, intrinsic_fns))
-                @timeit_debug to "device runtime library" begin
-                    linker = CUDAdrv.CuLink(jit_options)
-                    CUDAdrv.add_file!(linker, libcudadevrt[], CUDAdrv.JIT_INPUT_LIBRARY)
-                    CUDAdrv.add_data!(linker, kernel_fn, asm)
-                    image = CUDAdrv.complete(linker)
-                end
-            end
-
-            # JIT into an executable kernel object
-            mod = CuModule(image, jit_options)
-            fun = CuFunction(mod, kernel_fn)
-            kernel = HostKernel{f,tt}(ctx, mod, fun)
-
-            create_exceptions!(mod)
-
-            compilecache[key] = kernel
-            stop = time_ns()
-            @debug begin
-                ver = version(kernel)
-                mem = memory(kernel)
-                reg = registers(kernel)
-                fn = something(name, nameof(f))
-                """Compiled $fn($(join(tt.parameters, ", "))) to PTX $(ver.ptx) for SM $(ver.binary) in $(round((time_ns() - start) / 1000000; digits=2)) ms.
-                   Kernel uses $reg registers, and $(Base.format_bytes(mem.local)) local, $(Base.format_bytes(mem.shared)) shared, and $(Base.format_bytes(mem.constant)) constant memory."""
-            end
-        end
-
-        return compilecache[key]::HostKernel{f,tt}
-    end
+@generated function cufunction(f::Core.Function, tt::Type=Tuple{}; kwargs...)
+    # generated function that crafts a custom code info to call the actual cufunction impl.
+    # this gives us the flexibility to insert manual back edges for automatic recompilation.
+    tt = tt.parameters[1]
+
+    # get a hold of the method and code info of the kernel function
+    sig = Tuple{f, tt.parameters...}
+    mthds = _methods_by_ftype(sig, -1, typemax(UInt))
+    length(mthds) == 1 || return (:(throw(MethodError(f,tt))))
+    mtypes, msp, m = mthds[1]
+    mi = ccall(:jl_specializations_get_linfo, Ref{MethodInstance}, (Any, Any, Any), m, mtypes, msp)
+    mi.def.isva && return :(error("varargs kernel methods are not supported"))
+    ci = retrieve_code_info(mi)
+    @assert isa(ci, CodeInfo)
+
+    # generate a unique id to represent this specialization
+    global specialization_counter
+    id = (specialization_counter += 1)
+    # TODO: save the mi/ci here (or embed it in the AST to pass to cufunction)
```

Comment:
Yeah probably just pass it to cufunction along with `f, tt`

Reply:
That results in a gcframe in the calling code though. Maybe that's a micro optimization, but it would be nice to get as clean as possible kernel launching code.